src/gas_metering/backend.rs

```diff
@@ -0,0 +1,114 @@
+//! Provides backends for the gas metering instrumentation
+use parity_wasm::{builder::FunctionDefinition, elements};
+
+/// Implementation details of the specific method of the gas metering.
+pub enum GasMeter {
+	/// Gas metering with an external function.
+	External {
+		/// Name of the module to import the gas function from.
+		module: &'static str,
+		/// Name of the external gas function to be imported.
+		function: &'static str,
+	},
+	/// Gas metering with a local function and a mutable global.
+	Internal {
+		/// Name of the mutable global to be exported.
+		global: &'static str,
+		/// Definition of the local gas counting function to be injected.
+		function: FunctionDefinition,
+	},
+}
+
+/// Under the hood part of the gas metering mechanics.
+pub trait Backend {
+	/// Provides the gas metering implementation details.  
+	fn gas_meter(self, module: elements::Module) -> GasMeter;
```

Comment:
```suggestion
	fn gas_meter(self, module: &elements::Module) -> GasMeter;
```

I see no reason why this should take the module by value.